### PR TITLE
fix(pdb): Enable running cloud-init under pdb

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message

```
fix(pdb): Enable running cloud-init under pdb

Cloud-init closes stdin on startup, which breaks interactively running cloud-init under pdb.

Leave stdin open if it is connected to a tty, which fixes pdb use.


File "/usr/lib/python3.12/bdb.py", line 90, in trace_dispatch
  return self.dispatch_line(frame)
         ^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/lib/python3.12/bdb.py", line 115, in dispatch_line
    if self.quitting: raise BdbQuit
                      ^^^^^^^^^^^^^
bdb.BdbQuit
------------------------------------------------------------
The program exited via sys.exit(). Exit status: 1
```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by CalvoM** [comment]
Thanks for this. I ran into this issue when trying to debug some issues.

**Review by CalvoM — APPROVED** [review]
LGTM.

File: cloudinit/cmd/main.py [hunk 1]
```diff
@@ -328,9 +328,12 @@ def main_init(name, args):
...
     errfmt = None
     try:
-        early_logs.append((logging.DEBUG, "Closing stdin."))
-        util.close_stdin()
-        (outfmt, errfmt) = util.fixup_output(init.cfg, name)
+        if not os.isatty(sys.stdin.fileno()):
+            early_logs.append((logging.DEBUG, "Closing stdin."))
+            util.close_stdin()
+        else:
+            LOG.warning("Not closing stdin, stdin is a tty.")
```
Comment (TheRealFalcon) [inline cloudinit/cmd/main.py:335]: Is there a reason for this to be a warning? I think it makes more sense as a debug.

File: cloudinit/cmd/main.py [hunk 2]
```diff
@@ -328,9 +328,12 @@ def main_init(name, args):
     outfmt = None
     errfmt = None
     try:
-        early_logs.append((logging.DEBUG, "Closing stdin."))
-        util.close_stdin()
-        (outfmt, errfmt) = util.fixup_output(init.cfg, name)
+        if not os.isatty(sys.stdin.fileno()):
```
Comment (TheRealFalcon) [inline cloudinit/cmd/main.py:331]: Nit: These 5 lines are duplicated a few other times. Any reason we can't move them all into `util.close_stdin()` so each call site is just the single call to the function?

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
I think we should change the log level. The rest of the comments are nits.

Nit: Think there's any need to keep the `_CLOUD_INIT_SAVE_STDIN` env var? If its only use was for debugging, I don't see a reason to keep it around.

File: cloudinit/cmd/main.py (hunk 2 above)
Comment (holmanb) [reply cloudinit/cmd/main.py]: I also bumped this function over to `cloudinit/cmd/main.py` since it isn't used anywhere else

**Comment by holmanb** [comment]
@TheRealFalcon @CalvoM Thanks for the reviews! I think I've addressed all of the comments, requesting re-review.

**Review by TheRealFalcon — APPROVED** [review]
LGTM!

**Pull request merged**